cirq/circuits/expectation_value.py

```diff
@@ -141,90 +154,113 @@ def expectation_from_sampling(circuit: Circuit,
 
         return expectation
 
-    if quadratic_z:
 
-        # Check whether circuit already has measurement gates:
-        # only applicable if calculating expectation by measurement
+def _expectation_from_sampling_assuming_quadratic(
+        circuit: Circuit,
+        operator: Dict[PauliString, float],
+        n_samples: int):
+    """
+    Helper function that calculates expectation value if operator is
+    quadratic and contains only Pauli Z operators.
+
+    Args:
+        circuit: circ.Circuit type which prepares the state we are interested
+                 in calculating the expectation value of operator
+        operator: Operator input as a dictionary where keys are
+                    cirq.ops.PauliString type and values are coefficients
+                  example:
+                  paulistring = cirq.PauliString(qubit_pauli_map=
+                    {qubits[0]:cirq.Pauli.X, qubits[1]: (cirq.Pauli.Z)})
+                  paulistring2 = cirq.PauliString(qubit_pauli_map=
+                    {qubits[0]:cirq.Pauli.Z, qubits[2]: (cirq.Pauli.Y)})
+                  operator = {paulistring: 1.234, paulistring2: -5.678}
+        n_samples: number of runs for sampling
+
+    Returns:
+        expectation value of operator argument
+    """
 
-        last_moment_index = len(circuit) - 1
-        last_operations = circuit[last_moment_index].operations
+    qubits = list(circuit.all_qubits())
+    sim = XmonSimulator()
 
-        measurement_qubits = list(qubits)
-        indices_measurement = list(range(len(qubits)))
+    # Check whether circuit already has measurement gates:
+    last_moment_index = len(circuit) - 1
+    last_operations = circuit[last_moment_index].operations
```

Comment:
Measurements aren't guaranteed to be in the final moment of a circuit. They may also appear in the middle of a circuit, though the simulation methods will appropriately raise if that is the case.

Reply:
I modified the checks to account for this.